### ISSUE TITLE
I removed the gui and made it into a cli script (and some other changes)

### ISSUE THREAD

**Issue by kalasbralla** [rupeshs/fastsdcpu#16]
**$ python main_cli.py -h:**
```bash
› python main_cli.py -h
usage: main_cli.py [-h] --prompt PROMPT
                   [--guidance-scale GUIDANCE_SCALE]
                   [--img-width IMG_WIDTH]
                   [--img-height IMG_HEIGHT]
                   [--inference-steps INFERENCE_STEPS]
                   [--lcm-model-id LCM_MODEL_ID]
                   [--use-openvino] [--use-seed]
                   [--use-safety-checker] [--show-image]

Image generation script.

options:
  -h, --help            show this help message and exit
  --prompt PROMPT       Image prompt.
  --guidance-scale GUIDANCE_SCALE
                        Guidance scale.
  --img-width IMG_WIDTH
                        Image width.
  --img-height IMG_HEIGHT
                        Image height.
  --inference-steps INFERENCE_STEPS
                        Number of inference steps.
  --lcm-model-id LCM_MODEL_ID
                        LCM model ID.
  --use-openvino        Use OpenVINO.
  --use-seed            Use seed.
  --use-safety-checker  Use safety checker.
  --show-image          Show generated image.
```

### Example Usages

1. **Generate an image using default values from `config.json`:**
   ```bash
   python main_cli.py --prompt "Beautiful sunset over the mountains"
   ```

2. **Overriding the image dimensions specified in `config.json`:**
   ```bash
   python main_cli.py --prompt "Serenity in the forest" --img-width 256 --img-height 256
   ```

3. **Use OpenVINO for the generation and display the image post-generation:**
   ```bash
   python main_cli.py --prompt "Majestic view of the Milky Way" --use-openvino --show-image
   ```

4. **Specify guidance scale and inference steps:**
   ```bash
   python main_cli.py --prompt "An ancient temple in the rainforest" --guidance-scale 0.8 --inference-steps 25
   ```

5. **Use a custom seed for reproducibility:**
   ```bash
   python main_cli.py --prompt "Crystal clear waters of a hidden lagoon" --use-seed
   ```

### Configuration and Overriding

The `config.json` file is a convenient way to set default values for various parameters. Each parameter in this file can be overridden by command-line flags. This means that even if you set a value in `config.json`, you can use a flag to specify a different value for that parameter when you run the script.

For instance, if `config.json` has `"img_width": 512`, you can override it with:

```
python main_cli.py --prompt "Your Prompt" --img-width 256
```

This provides flexibility in that you don't need to constantly modify the `config.json` file when you want to change certain parameters. The command-line flags will always take precedence.

### Transition to CLI

The transition from a PyQt5 GUI to a CLI-based approach was made to simplify the user experience, especially for those who prefer command-line operations. The added benefit of using `config.json` means users don't have to remember all the flags or their specific values every time they run the script; they can set their most frequently used parameters in the configuration file and then occasionally override them with flags as needed.

### Image Results and Metadata

Every time you generate an image, the tool not only saves the image but also a corresponding JSON file in the results directory. This JSON file details all the parameters used during the image generation process, ensuring you can keep track of or reproduce specific results.

If the `show_image` parameter is set to `true` in `config.json` (or if you use the `--show-image` flag), the tool will also display the generated image immediately after it's created.

**Example output:**

![image](https://github.com/rupeshs/fastsdcpu/assets/53936033/75068f42-1c1c-4918-8541-adb57a4c61f6)

**main_cli.py:**

```python
from diffusers import DiffusionPipeline
from time import time
import sys
import os
import warnings
import torch
import argparse
import json
import logging


from uuid import uuid4
from openvino.lcm_ov_pipeline import OVLatentConsistencyModelPipeline
from openvino.lcm_scheduler import LCMScheduler

# Suppress FutureWarnings
warnings.simplefilter(action='ignore', category=FutureWarning)

# Setup logging
logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

RESULTS_DIRECTORY = "results"

if not os.path.exists(RESULTS_DIRECTORY):
    os.makedirs(RESULTS_DIRECTORY)


def set_seed(use_seed=True, seed_value=None):
    if use_seed:
        if seed_value is None:
            seed_value = torch.randint(0, 1 << 31, (1,)).item()
        torch.manual_seed(seed_value)
        logger.info(f"Using seed value: {seed_value}")
        return seed_value
    else:
        return None


def load_config(config_file='config.json'):
    defaults = {
        'show_image': False,
        'guidance_scale': None,
        'img_width': None,
        'img_height': None,
        'inference_steps': None,
        'lcm_model_id': None,
        'use_openvino': None,
        'use_seed': None,
        'use_safety_checker': None
    }
    try:
        with open(config_file, 'r') as file:
            config = json.load(file)
            return {**defaults, **config}
    except FileNotFoundError:
        logger.error(f"Configuration file {config_file} not found.")
        sys.exit(1)
    except json.JSONDecodeError:
        logger.error(
            f"Failed to parse configuration file {config_file}. Ensure it contains valid JSON.")
        sys.exit(1)


config = load_config()
# default to False if not in config
show_image = config.get('show_image', False)


def get_results_path():
    app_dir = os.path.dirname(__file__)
    config_path = os.path.join(app_dir, RESULTS_DIRECTORY)
    return config_path


def generate_image(prompt, guidance_scale, img_width, img_height, num_inference_steps, use_openvino, model_id, use_seed, seed_value, safety_checker):
    output_path = get_results_path()

    if use_openvino:
        logger.info("Using OpenVINO for image generation.")
        scheduler = LCMScheduler.from_pretrained(
            model_id, subfolder="scheduler")
        pipeline = OVLatentConsistencyModelPipeline.from_pretrained(
            "deinferno/LCM_Dreamshaper_v7-openvino",
            scheduler=scheduler,
            compile=False,
        )
    else:
        logger.info("Not using OpenVINO for image generation.")
        pipeline = DiffusionPipeline.from_pretrained(
            model_id,
            custom_pipeline="latent_consistency_txt2img",
            custom_revision="main",
        )
        pipeline.to(torch_device="cpu", torch_dtype=torch.float32)

    logger.info(f"Prompt : {prompt}")
    logger.info(f"Resolution : {img_width} x {img_height}")
    logger.info(f"Guidance Scale : {guidance_scale}")
    logger.info(f"Inference_steps  : {num_inference_steps}")

    if use_seed:
        logger.info(f"Seed: {seed_value}")

    tick = time()

    if not safety_checker:
        pipeline.safety_checker = None

    images = pipeline(
        prompt=prompt,
        num_inference_steps=num_inference_steps,
        guidance_scale=guidance_scale,
        lcm_origin_steps=50,
        width=img_width,
        height=img_height,
        output_type="pil",
    ).images

    elapsed = time() - tick
    logger.info(f"Elapsed time : {elapsed:.2f} sec")
    image_id = uuid4()

    # Save the image
    image_path = os.path.join(output_path, f"{image_id}.png")
    images[0].save(image_path)
    logger.info(f"Image {image_id}.png saved")

    # Show image if true
    if show_image:
        logger.info(f"show_image: true - Displaying {image_id}.png")
        images[0].show()

    # Save the parameters used in a JSON file with the same UUID
    with open(os.path.join(output_path, f"{image_id}.json"), 'w') as json_file:
        json.dump({
            'prompt': prompt,
            'guidance_scale': guidance_scale,
            'img_width': img_width,
            'img_height': img_height,
            'num_inference_steps': num_inference_steps,
            'use_openvino': use_openvino,
            'model_id': model_id,
            'use_seed': use_seed,
            'seed_value': seed_value,
            'safety_checker': safety_checker,
            'elapsed_time_sec': elapsed
        }, json_file, indent=4)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Image generation script.')

    parser.add_argument('--prompt', type=str, required=True,
                        help='Image prompt.')
    parser.add_argument('--guidance-scale', type=float, default=config['guidance_scale'],
                        help='Guidance scale.')
    parser.add_argument('--img-width', type=int, default=config['img_width'],
                        help='Image width.')
    parser.add_argument('--img-height', type=int, default=config['img_height'],
                        help='Image height.')
    parser.add_argument('--inference-steps', type=int, default=config['inference_steps'],
                        help='Number of inference steps.')
    parser.add_argument('--lcm-model-id', type=str, default=config['lcm_model_id'],
                        help='LCM model ID.')
    parser.add_argument('--use-openvino', action='store_true', default=config['use_openvino'],
                        help='Use OpenVINO.')
    parser.add_argument('--use-seed', action='store_true', default=config['use_seed'],
                        help='Use seed.')
    parser.add_argument('--use-safety-checker', action='store_true', default=config['use_safety_checker'],
                        help='Use safety checker.')
    parser.add_argument('--show-image', action='store_true', default=config.get('show_image', False),
                        help='Show generated image.')
    parser.add_argument('--seed', type=int, default=None,
                        help='Specific seed for reproducibility. If not provided, a random seed will be used.')

    args = parser.parse_args()

    # Set the seed here and use the returned value throughout the script
    seed_value = set_seed(args.use_seed, args.seed)

    generate_image(args.prompt, args.guidance_scale, args.img_width, args.img_height, args.inference_steps,
                   args.use_openvino, args.lcm_model_id, args.use_seed, seed_value, args.use_safety_checker)

```

**config.json:**
```json
{
  "guidance_scale": 4.0,
  "img_width": 256,
  "img_height": 256,
  "inference_steps": 5,
  "use_openvino": true,
  "lcm_model_id": "SimianLuo/LCM_Dreamshaper_v7",
  "use_seed": true,
  "use_safety_checker": false,
  "show_image": true
}
```
I've tested the script on Linux, and it seems to be functioning as expected. Any feedback or suggestions for improvement are appreciated!

**Comment by Accretence** [comment]
It would be very interesting to build this into a project deployable to services like Heroku and Railway where can use it using REST APIs.

**Comment by kalasbralla** [comment]
I've done small updates to the script and some improvements. The seeding seems to work much better now and the results are reproducible.

**Comment by rupeshs** [comment]
CLI added in the release 1.0.0 beta 7
